cluster/examples/kubernetes/ceph/cluster.yaml

```diff
@@ -49,6 +55,58 @@ subjects:
   name: rook-ceph-cluster
   namespace: rook-ceph
 ---
+# Extra powers for the ceph-mgr daemon, as it may run a k8s client
+# for managing Rook from within.
+kind: ClusterRoleBinding
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-ceph-mgr
+  namespace: rook-ceph
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: rook-ceph-mgr
+subjects:
+- kind: ServiceAccount
+  name: rook-ceph-mgr
+  namespace: rook-ceph
+---
+# TODO: scope this down from a ClusterRole to a Role
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1beta1
+metadata:
+  name: rook-ceph-mgr
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  - services
+  - nodes
+  - nodes/proxy
+  - configmaps
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - batch
```

Comment:
Anything that gives full access makes me wonder if there is a way we can use the operator to perform the task. For example, if you need to launch a job to repair a filesystem, it seems like the operator should be able to detect that condition and start the operation. Could the operator query the mgr for health and trigger such operations? Or are we saying that we expect the mgr modules to handle certain health management of the cluster? I'm still not understanding the line in the sand between the operator and mgr.

Reply:
> it seems like the operator should be able to detect that condition and start the operation

This is a case where we're just not there yet in terms of autonomic operation -- diagnosing and correcting filesystem corruptions is not something Ceph (or Rook) can just do automatically.  Even when this gets more automated than it is today, we wouldn't necessarily want the system to do it without operator intervention, as it's all about dealing with unexpected circumstances, where something unknown has gone wrong, and any repair activity needs to tread carefully.

It's already possible for ceph-mgr to internally run tools (well, with a few tweaks to linkage to let ceph-mgr get at them) -- the only reason the orchestrator enters the discussion is the need to scale out certain operations (brute force scans of data objects, especially) across a number of workers.  It makes sense to me to keep the Ceph-specific knowledge (e.g. the painfully detailed knowledge of how to use cephfs-data-scan) inside Ceph, and have the orchestrator provide just the scaling out part -- that way we don't have multiple orchestrators trying to understand how to do CephFS repair.

Having the orchestrator just provide the scale-out part doesn't have to mean consuming kubernetes APIs directly.  We could have Rook provide e.g. a slightly sandboxed batch API that only permitted running the Ceph image.  I'm not sure how in-scope that is for Rook: it's a mission change from "run storage system" to "provide a worker resource for the storage system".